### PR TITLE
Changes to ensure CompiledPackages contains contract key template lambda definitions

### PR DESCRIPTION
**Pull request by carlpulley-da**
This PR enables work started under https://github.com/digital-asset/daml/pull/14609 to be completed.

Compiler changes:
- add in `ContractKeyWithMaintainersDefRef`
- ensure `compileModule` compiles contract key lambdas for templates (and so allow contract keys to be extracted from contract instances)

CHANGELOG_BEGIN
CHANGELOG_END

### CODE REVIEW

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala [hunk 1]
```diff
@@ -804,6 +805,22 @@ private[lf] final class Compiler(
...
+  private[this] def compileContractKeyWithMaintainers(
+      tmplId: Identifier,
+      tmpl: Template,
+      tmplKey: TemplateKey,
+  ): (t.SDefinitionRef, SDefinition) = {
+    // compile a template with key into:
+    // ContractKeyWithMaintainersDefRef(tmplId) = \ <tmplArg> <token> ->
+    //   let <key> = tmplKey.body(<tmplArg>)
+    //   in { key = <key> ; maintainers = [tmplKey.maintainers] <key> }
+    topLevelFunction2(t.ContractKeyWithMaintainersDefRef(tmplId)) { (tmplArg, _, env) =>
```
Comment (nickchapman-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala:817]: Can we gain confidence in the definition of `ContractKeyWithMaintainersDefRef` by using it in `compileToCachedContract` ?

**Review by nickchapman-da — APPROVED** [review]

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala (hunk 1 above)
Comment (carlpulley-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala:817]: - `ToCachedContractDefRef` captures the lambda with the contract (or template argument) and the key instance as the lambdas 2 bound variables.

- `ContractKeyWithMaintainersDefRef` captures the lambda with the contract (or template argument) as the lambdas sole bound variable.

Whether its useful to rationalise these 2 behaviours is something I can look into as I build upon this code (I will certainly need to build cached contracts in my next steps here).

Comment (carlpulley-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala:817]: This said, maybe some simple happy path testing would be a good idea here - if only to demo how this code is intended to be used.

**Review by akrmn — APPROVED** [review]
LGTM

**Review by remyhaemmerle-da — COMMENTED** [review]
:+1:

**Pull request merged**